File: kitchen_table/rendering.py
from __future__ import annotations
import os
from pyray import *
import raylib as rl
from kitchen_table.models import Card, Stack, Table_State
from kitchen_table.config import tweak

# Background shader
_background_shader = None
_background_time_loc = -1
_background_resolution_loc = -1
_background_turn_loc = -1
_background_turn_value = 0.0


def _load_background_shader():
    global _background_shader, _background_time_loc, _background_resolution_loc, _background_turn_loc
    shader_path = os.path.join(os.path.dirname(__file__), "background.fs")
    with open(shader_path) as f:
        fs_code = f.read()
    _background_shader = load_shader_from_memory(rl.ffi.NULL, fs_code.encode())
    _background_time_loc = get_shader_location(_background_shader, "u_time")
    _background_resolution_loc = get_shader_location(_background_shader, "u_resolution")
    _background_turn_loc = get_shader_location(_background_shader, "u_turn")


def draw_background(turn: float = 0.0):
    global _background_shader, _background_turn_value
    if _background_shader is None:
        _load_background_shader()

    # Smoothly interpolate toward the target turn value.
    dt = get_frame_time()
    speed = 3.0
    _background_turn_value += (turn - _background_turn_value) * min(dt * speed, 1.0)

    t = rl.ffi.new("float *", get_time())
    rl.SetShaderValue(_background_shader, _background_time_loc, t,
                      rl.SHADER_UNIFORM_FLOAT)

    res = rl.ffi.new("float[2]", [get_screen_width(), get_screen_height()])
    rl.SetShaderValue(_background_shader, _background_resolution_loc, res,
                      rl.SHADER_UNIFORM_VEC2)

    turn_val = rl.ffi.new("float *", _background_turn_value)
    rl.SetShaderValue(_background_shader, _background_turn_loc, turn_val,
                      rl.SHADER_UNIFORM_FLOAT)

    begin_shader_mode(_background_shader)
    draw_rectangle(0, 0, get_screen_width(), get_screen_height(), WHITE)
    end_shader_mode()


# Font: loaded lazily on first use after init_window.
_font: Font | None = None


def get_font() -> Font:
    """Return the configured font, loading it on first call."""
    global _font
    if _font is None:
        font_path = tweak.get("font_path", "")
        if font_path and os.path.exists(font_path):
            _font = load_font_ex(font_path.encode(), tweak["font_load_size"], None, 0)
        else:
            _font = get_font_default()
        # Bilinear filtering prevents jagged edges when scaling to any size.
        set_texture_filter(_font.texture, TextureFilter.TEXTURE_FILTER_BILINEAR)
    return _font


def render_text(text: str, x: int, y: int, size: int, color: Color) -> None:
    """Draw text using the configured font."""
    draw_text_ex(get_font(), text, Vector2(x, y), float(size), tweak["font_spacing"], color)


def text_width(text: str, size: int) -> int:
    """Measure text width using the configured font."""
    return int(measure_text_ex(get_font(), text, float(size), tweak["font_spacing"]).x)


# Texture cache to avoid reloading images
_texture_cache: dict[str, Texture2D] = {}


def get_texture(image_path: str) -> Texture2D | None:
    """Load and cache a texture from disk."""
    if image_path in _texture_cache:
        return _texture_cache[image_path]

    if not file_exists(image_path.encode()):
        return None

    texture = load_texture(image_path.encode())
    _texture_cache[image_path] = texture
    return texture


_rounded_texture_cache: dict[str, Texture2D] = {}


def get_rounded_texture(image_path: str) -> Texture2D | None:
    """Load a texture with rounded corners applied, using cache."""
    if image_path in _rounded_texture_cache:
        return _rounded_texture_cache[image_path]

    if not file_exists(image_path.encode()):
        return None

    w = tweak["card_width"]
    h = tweak["card_height"]
    r = tweak["card_corner_radius"]

    # Load image at original resolution (GPU scales when drawing)
    image = load_image(image_path.encode())
    iw = image.width
    ih = image.height

    # Scale corner radius to match image resolution
    sr = int(r * min(iw / w, ih / h))

    # Create rounded rectangle mask at image resolution
    mask = gen_image_color(iw, ih, Color(0, 0, 0, 0))
    image_draw_rectangle(mask, sr, 0, iw - 2 * sr, ih, WHITE)
    image_draw_rectangle(mask, 0, sr, iw, ih - 2 * sr, WHITE)
    image_draw_circle(mask, sr, sr, sr, WHITE)
    image_draw_circle(mask, iw - sr, sr, sr, WHITE)
    image_draw_circle(mask, sr, ih - sr, sr, WHITE)
    image_draw_circle(mask, iw - sr, ih - sr, sr, WHITE)

    # Apply mask and convert to GPU texture
    image_alpha_mask(image, mask)
    texture = load_texture_from_image(image)

    unload_image(image)
    unload_image(mask)

    _rounded_texture_cache[image_path] = texture
    return texture


def color_from_tuple(c: tuple) -> Color:
    """Convert RGBA tuple to raylib Color."""
    return Color(c[0], c[1], c[2], c[3])


def draw_card_back() -> None:
    x = 0
    y = 0
    """Draw a face-down card."""
    w = tweak["card_width"]
    h = tweak["card_height"]
    r = tweak["card_corner_radius"]

    # Card background
    draw_rectangle_rounded(
        Rectangle(x, y, w, h), r / min(w, h), 8,
        color_from_tuple(tweak["card_back"])
    )

    # Simple pattern on back
    pattern_color = color_from_tuple(tweak["card_back_pattern"])
    margin = 15
    draw_rectangle_rounded(
        Rectangle(x + margin, y + margin, w - 2*margin, h - 2*margin),
        r / min(w, h), 8, pattern_color
    )

    # Border
    draw_rectangle_rounded_lines_ex(
        Rectangle(x, y, w, h), r / min(w, h), 8, 2,
        color_from_tuple(tweak["card_border"])
    )


def draw_card_content(card: Card, face_up: bool) -> None:
    """Draw card content at origin (used for rotation)."""
    if not face_up:
        draw_card_back()
        return

    x = 0
    y = 0
    w = tweak["card_width"]
    h = tweak["card_height"]
    r = tweak["card_corner_radius"]
    padding = tweak["card_padding"]

    # Card background (image or solid color)
    texture = None
    if card.image_path:
        texture = get_rounded_texture(card.image_path)

    if texture:
        # Draw image scaled to fit card
        source_rect = Rectangle(0, 0, texture.width, texture.height)
        dest_rect = Rectangle(x, y, w, h)
        draw_texture_pro(texture, source_rect, dest_rect, Vector2(0, 0), 0, WHITE)
    else:
        # Fallback to solid color background
        draw_rectangle_rounded(
            Rectangle(x, y, w, h), r / min(w, h), 8,
            color_from_tuple(tweak["card_background"])
        )

    if card.draw_callback:
        card.draw_callback(card)

    return
    # Border
    draw_rectangle_rounded_lines_ex(
        Rectangle(x, y, w, h), r / min(w, h), 8, 2,
        color_from_tuple(tweak["card_border"])
    )

    # Title with background for readability when image is present
    title_size = tweak["title_font_size"]
    title_text = card.title
    title_w = text_width(title_text, title_size)
    if texture:
        # Draw semi-transparent background behind title
        draw_rectangle(
            int(x + padding - 2),
            int(y + padding - 2),
            title_w + 4,
            title_size + 4,
            Color(0, 0, 0, 180)
        )
    render_text(
        title_text,
        int(x + padding),
        int(y + padding),
        title_size,
        color_from_tuple(tweak["card_title_color"])
    )

    # Description (with simple word wrapping)
    if card.description:
        desc_size = tweak["description_font_size"]
        desc_y = y + padding + title_size + 10
        max_width = w - 2 * padding

        # Simple word wrapping
        words = card.description.split()
        lines = []
        current_line = ""
        for word in words:
            test_line = current_line + " " + word if current_line else word
            line_w = text_width(test_line, desc_size)
            if line_w <= max_width:
                current_line = test_line
            else:
                if current_line:
                    lines.append(current_line)
                current_line = word
        if current_line:
            lines.append(current_line)

        # Draw background for description if image is present
        if texture and lines:
            total_height = len(lines) * (desc_size + 2)
            draw_rectangle(
                int(x + padding - 2),
                int(desc_y - 2),
                int(max_width + 4),
                int(total_height + 4),
                Color(0, 0, 0, 180)
            )

        for i, line in enumerate(lines):
            render_text(
                line,
                int(x + padding),
                int(desc_y + i * (desc_size + 2)),
                desc_size,
                color_from_tuple(tweak["card_description_color"])
            )


def draw_card(card: Card, face_up: bool = True) -> None:
    """Draw a single card at its position, with rotation support."""
    w = tweak["card_width"]
    h = tweak["card_height"]

    # Apply rotation around center
    # Calculate center of card
    if card.rotation == 0:
        cx = card.x
        cy = card.y
        rl_push_matrix()
        rl_translatef(cx, cy, 0)
        draw_card_content(card, face_up)
        rl_pop_matrix()
    else:
        # Calculate center of card
        cx = card.x + w / 2
        cy = card.y + h / 2
        # Apply rotation around center
        rl_push_matrix()
        rl_translatef(cx, cy, 0)
        rl_rotatef(card.rotation, 0, 0, 1)
        rl_translatef(-w/2, -h/2, 0)
        draw_card_content(card, face_up)
        rl_pop_matrix()


def draw_stack(stack: Stack, state: Table_State) -> None:
    """Draw all cards in a stack."""
    for card_id in stack.cards:
        card = state.cards[card_id]
        draw_card(card, face_up=stack.face_up)


def draw_stack_placeholder(stack: Stack) -> None:
    """Draw an empty stack placeholder with label."""
    w = tweak["card_width"]
    h = tweak["card_height"]
    r = tweak["card_corner_radius"]

    # Dashed outline placeholder
    draw_rectangle_rounded_lines_ex(
        stack.rect, r / min(w, h), 8, 1,
        Color(100, 100, 100, 100)
    )

    # Label
    label = stack.name
    label_w = text_width(label, 14)
    render_text(
        label,
        int(stack.rect.x + (w - label_w) / 2),
        int(stack.rect.y + h / 2 - 7),
        14,
        Color(100, 100, 100, 150)
    )


def animate(cards, state, dt: float = 0.1) -> None:
    # Interpolate card positions
    for r_card, t_card in zip(cards, state.cards):
        old_pos_x = r_card.x
        r_card.x = r_card.x * (1 - dt) + t_card.x * dt
        r_card.y = r_card.y * (1 - dt) + t_card.y * dt
        v_x = r_card.x - old_pos_x
        r_card.rotation = r_card.rotation * (1 - dt) + t_card.rotation * dt
        r_card.rotation += v_x * 0.1

    selected_card_id = state.drag_state.card_id
    if selected_card_id >= 0:
        cards[selected_card_id].x = state.cards[selected_card_id].x
        cards[selected_card_id].y = state.cards[selected_card_id].y

def draw_zoomed_card(card: Card, face_up: bool) -> None:
    """Draw a card fullscreen as a zoom preview."""
    screen_w = get_screen_width()
    screen_h = get_screen_height()

    # Dim background
    draw_rectangle(0, 0, screen_w, screen_h, Color(0, 0, 0, 160))

    # Scale card to fit screen with some margin
    card_w = tweak["card_width"]
    card_h = tweak["card_height"]
    margin = 40
    scale = min((screen_w - 2 * margin) / card_w, (screen_h - 2 * margin) / card_h)

    # Center on screen
    cx = (screen_w - card_w * scale) / 2
    cy = (screen_h - card_h * scale) / 2

    rl_push_matrix()
    rl_translatef(cx, cy, 0)
    rl_scalef(scale, scale, 1)
    draw_card_content(card, face_up=face_up)
    rl_pop_matrix()


from copy import deepcopy

def draw_table(table_state: Table_State) -> None:
    if table_state.animated_cards is None:
        table_state.animated_cards = deepcopy(table_state.cards)

    animate(table_state.animated_cards, table_state)

    drag = table_state.drag_state

    # Draw stack placeholders for empty stacks
    if drag.current_stack != -1:
        stack = table_state.stacks[drag.current_stack]
        draw_stack_placeholder(stack)

    # Draw stacks,
    for stack in sorted(table_state.stacks, key=lambda s: s.depth):
        for card_id in stack.cards:
            card = table_state.animated_cards[card_id]
            if table_state.drag_state.card_id == card_id:
                continue
            draw_card(card, face_up=stack.face_up)

    # Draw loose cards
    for card_id in table_state.loose_cards:
        card = table_state.animated_cards[card_id]
        draw_card(card, face_up=True)

    if table_state.drag_state.card_id >= 0:
        card = table_state.animated_cards[table_state.drag_state.card_id]
        face_up = table_state.stacks[table_state.drag_state.original_stack].face_up
        draw_card(card, face_up=face_up)

    if table_state.draw_callback is not None:
        table_state.draw_callback(table_state)

    # Draw zoomed card on top of everything
    if table_state.zoomed_card_id >= 0:
        face_up = True
        for stack in table_state.stacks:
            if table_state.zoomed_card_id in stack.cards:
                face_up = stack.face_up
        draw_zoomed_card(table_state.cards[table_state.zoomed_card_id], face_up=face_up)

File: kitchen_table_cpp/code/rendering.cpp
#include "rendering.h"

#include <algorithm>
#include <cmath>
#include <string>
#include <unordered_map>

#include "config.h"
#include "raylib.h"
#include "rlgl.h"  // for rlPushMatrix, rlPopMatrix, rlTranslatef, rlRotatef, rlScalef

// --- Static globals (lazy initialized) ---

static Shader s_background_shader = {0};
static int    s_bg_time_loc       = -1;
static int    s_bg_resolution_loc = -1;
static int    s_bg_turn_loc       = -1;
static float  s_bg_turn_value     = 0.0f;

static Font s_font        = {0};
static bool s_font_loaded = false;

static std::unordered_map<std::string, Texture2D> s_texture_cache;
static std::unordered_map<std::string, Texture2D> s_rounded_texture_cache;


// --- Background shader loading ---

static void load_background_shader() {
  // Load the fragment shader source from disk.
  char* fs_code       = LoadFileText("kitchen_table_cpp/code/background.frag");
  s_background_shader = LoadShaderFromMemory(nullptr, fs_code);
  s_bg_time_loc       = GetShaderLocation(s_background_shader, "u_time");
  s_bg_resolution_loc = GetShaderLocation(s_background_shader, "u_resolution");
  s_bg_turn_loc       = GetShaderLocation(s_background_shader, "u_turn");
  UnloadFileText(fs_code);
}

// --- Font loading ---

static Font& get_font() {
  if (!s_font_loaded) {
    const char* font_path = kt::FONT_PATH;
    if (FileExists(font_path)) {
      s_font = LoadFontEx(font_path, kt::FONT_LOAD_SIZE, nullptr, 0);
    } else {
      s_font = GetFontDefault();
    }
    // Bilinear filtering prevents jagged edges when scaling to any size.
    SetTextureFilter(s_font.texture, TEXTURE_FILTER_BILINEAR);
    s_font_loaded = true;
  }
  return s_font;
}

// --- Texture cache helpers ---

static Texture2D* get_texture(const std::string& image_path) {
  auto it = s_texture_cache.find(image_path);
  if (it != s_texture_cache.end()) {
    return &it->second;
  }

  if (!FileExists(image_path.c_str())) {
    return nullptr;
  }

  Texture2D tex               = LoadTexture(image_path.c_str());
  s_texture_cache[image_path] = tex;
  return &s_texture_cache[image_path];
}

static Texture2D* get_rounded_texture(const std::string& image_path) {
  auto it = s_rounded_texture_cache.find(image_path);
  if (it != s_rounded_texture_cache.end()) {
    return &it->second;
  }

  if (!FileExists(image_path.c_str())) {
    return nullptr;
  }

  float w = (float)kt::CARD_WIDTH;
  float h = (float)kt::CARD_HEIGHT;
  float r = (float)kt::CARD_CORNER_RADIUS;

  // Load image at original resolution (GPU scales when drawing).
  Image image = LoadImage(image_path.c_str());
  int   iw    = image.width;
  int   ih    = image.height;

  // Scale corner radius to match image resolution.
  int sr = (int)(r * std::min((float)iw / w, (float)ih / h));

  // Create rounded rectangle mask at image resolution.
  Image mask = GenImageColor(iw, ih, Color{0, 0, 0, 0});
  ImageDrawRectangle(&mask, sr, 0, iw - 2 * sr, ih, WHITE);
  ImageDrawRectangle(&mask, 0, sr, iw, ih - 2 * sr, WHITE);
  ImageDrawCircle(&mask, sr, sr, sr, WHITE);
  ImageDrawCircle(&mask, iw - sr, sr, sr, WHITE);
  ImageDrawCircle(&mask, sr, ih - sr, sr, WHITE);
  ImageDrawCircle(&mask, iw - sr, ih - sr, sr, WHITE);

  // Apply mask and convert to GPU texture.
  ImageAlphaMask(&image, mask);
  Texture2D tex = LoadTextureFromImage(image);

  UnloadImage(image);
  UnloadImage(mask);

  s_rounded_texture_cache[image_path] = tex;
  return &s_rounded_texture_cache[image_path];
}

// --- Text rendering ---

void render_text(
  const std::string& text, float x, float y, int size, Color color
) {
  DrawTextEx(
    get_font(),
    text.c_str(),
    {(float)x, (float)y},
    (float)size,
    kt::FONT_SPACING,
    color
  );
}

int text_width(const std::string& text, int size) {
  Vector2 measured =
    MeasureTextEx(get_font(), text.c_str(), (float)size, kt::FONT_SPACING);
  return (int)measured.x;
}

// --- draw_background ---

void draw_background(float turn) {
  if (s_bg_time_loc == -1) {
    load_background_shader();
  }

  // Smoothly interpolate toward the target turn value.
  float dt    = GetFrameTime();
  float speed = 3.0f;
  s_bg_turn_value += (turn - s_bg_turn_value) * std::min(dt * speed, 1.0f);

  float t = (float)GetTime();
  SetShaderValue(s_background_shader, s_bg_time_loc, &t, SHADER_UNIFORM_FLOAT);

  float res[2] = {(float)GetScreenWidth(), (float)GetScreenHeight()};
  SetShaderValue(
    s_background_shader, s_bg_resolution_loc, res, SHADER_UNIFORM_VEC2
  );

  SetShaderValue(
    s_background_shader, s_bg_turn_loc, &s_bg_turn_value, SHADER_UNIFORM_FLOAT
  );

  BeginShaderMode(s_background_shader);
  DrawRectangle(0, 0, GetScreenWidth(), GetScreenHeight(), WHITE);
  EndShaderMode();
}

// --- draw_card_back ---

void draw_card_back() {
  float x = 0.0f;
  float y = 0.0f;
  float w = (float)kt::CARD_WIDTH;
  float h = (float)kt::CARD_HEIGHT;
  float r = (float)kt::CARD_CORNER_RADIUS;

  // KT_Card colors from kt namespace (matching config.py defaults).
  Color back_color    = {60, 80, 120, 255};
  Color pattern_color = {80, 100, 140, 255};
  Color border_color  = {80, 80, 80, 255};

  // KT_Card background.
  DrawRectangleRounded(
    Rectangle{x, y, w, h}, r / std::min(w, h), 8, back_color
  );

  // Simple pattern on back.
  float margin = 15.0f;
  DrawRectangleRounded(
    Rectangle{x + margin, y + margin, w - 2.0f * margin, h - 2.0f * margin},
    r / std::min(w, h),
    8,
    pattern_color
  );

  // Border.
  DrawRectangleRoundedLinesEx(
    Rectangle{x, y, w, h}, r / std::min(w, h), 8, 2.0f, border_color
  );
}

// --- draw_card_content ---

void draw_card_content(const Thing& card, bool face_up) {
  if (!face_up) {
    draw_card_back();
    return;
  }

  float x = 0.0f;
  float y = 0.0f;
  float w = (float)kt::CARD_WIDTH;
  float h = (float)kt::CARD_HEIGHT;
  float r = (float)kt::CARD_CORNER_RADIUS;

  // KT_Card background: image with rounded corners, or solid color fallback.
  Texture2D* texture = nullptr;
  if (!card.image_path.empty()) {
    texture = get_rounded_texture(card.image_path);
  }

  if (texture) {
    // Draw image scaled to fill the card.
    Rectangle source_rect = {
      0.0f, 0.0f, (float)texture->width, (float)texture->height
    };
    Rectangle dest_rect = {x, y, w, h};
    DrawTexturePro(
      *texture, source_rect, dest_rect, Vector2{0.0f, 0.0f}, 0.0f, WHITE
    );
  } else {
    // Fallback: solid color background.
    Color bg = {255, 255, 255, 255};
    DrawRectangleRounded(Rectangle{x, y, w, h}, r / std::min(w, h), 8, bg);
  }

  // Invoke the draw callback if set.
  if (card.draw_callback) {
    card.draw_callback(const_cast<Thing&>(card));
  }
}

// --- draw_card ---

void draw_card(const Thing& card, bool face_up) {
  float w = (float)kt::CARD_WIDTH;
  float h = (float)kt::CARD_HEIGHT;

  if (card.rotation == 0) {
    // No rotation: translate to card position and draw at origin.
    rlPushMatrix();
    rlTranslatef(card.x, card.y, 0.0f);
    draw_card_content(card, face_up);
    rlPopMatrix();
  } else {
    // Rotate around the center of the card.
    float cx = card.x + w / 2.0f;
    float cy = card.y + h / 2.0f;
    rlPushMatrix();
    rlTranslatef(cx, cy, 0.0f);
    rlRotatef((float)card.rotation, 0.0f, 0.0f, 1.0f);
    rlTranslatef(-w / 2.0f, -h / 2.0f, 0.0f);
    draw_card_content(card, face_up);
    rlPopMatrix();
  }
}

// --- draw_stack ---

void draw_stack(const Stack& stack, const Table_State& state) {
  // Draw each card that belongs to this stack.
  for (int card_id : stack.cards) {
    draw_card(state.cards[card_id], stack.face_up);
  }
}

// --- draw_stack_placeholder ---

void draw_stack_placeholder(const Stack& stack) {
  float w = (float)kt::CARD_WIDTH;
  float h = (float)kt::CARD_HEIGHT;
  float r = (float)kt::CARD_CORNER_RADIUS;

  // Dashed outline placeholder.
  DrawRectangleRoundedLinesEx(
    Rectangle{stack.rect.x, stack.rect.y, stack.rect.width, stack.rect.height},
    r / std::min(w, h),
    8,
    1.0f,
    Color{100, 100, 100, 100}
  );

  // Label centered in the placeholder.
  const std::string& label   = stack.name;
  int                label_w = text_width(label, 14);

  render_text(
    label,
    (int)(stack.rect.x + (w - (float)label_w) / 2.0f),
    (int)(stack.rect.y + h / 2.0f - 7.0f),
    14,
    Color{100, 100, 100, 150}
  );
}

// --- animate ---

void animate(std::vector<KT_Card>& cards, const Table_State& state, float dt) {
  int selected_card_id = state.drag_state.card_id;
  int n                = (int)cards.size();
  for (int i = 0; i < n; ++i) {
    KT_Card&       acard  = cards[i];
    const KT_Card& target = state.cards[i];

    float old_x = acard.x;
    acard.x     = acard.x * (1.0f - dt) + target.x * dt;
    acard.y     = acard.y * (1.0f - dt) + target.y * dt;
    float vx    = acard.x - old_x;
    acard.rotation =
      acard.rotation * (1.0f - dt) + target.rotation * dt + vx * 0.1f;
  }

  // Dragged card snaps immediately to the target position.
  if (selected_card_id >= 0 && selected_card_id < n) {
    KT_Card&       acard  = cards[selected_card_id];
    const KT_Card& target = state.cards[selected_card_id];
    acard.x            = target.x;
    acard.y            = target.y;
  }
}

// --- draw_zoomed_card ---

void draw_zoomed_card(const Thing& card, bool face_up) {
  int screen_w = GetScreenWidth();
  int screen_h = GetScreenHeight();

  // Dim background.
  DrawRectangle(0, 0, screen_w, screen_h, Color{0, 0, 0, 160});

  float card_w = (float)kt::CARD_WIDTH;
  float card_h = (float)kt::CARD_HEIGHT;
  float margin = 40.0f;

  // Scale card to fill screen with some margin.
  float scale = std::min(
    ((float)screen_w - 2.0f * margin) / card_w,
    ((float)screen_h - 2.0f * margin) / card_h
  );

  // Center on screen.
  float cx = ((float)screen_w - card_w * scale) / 2.0f;
  float cy = ((float)screen_h - card_h * scale) / 2.0f;

  rlPushMatrix();
  rlTranslatef(cx, cy, 0.0f);
  rlScalef(scale, scale, 1.0f);
  draw_card_content(card, face_up);
  rlPopMatrix();
}

// --- draw_table ---

void draw_table(Table_State& state) {
  // Initialize animated_cards on first call (copy of cards list).
  if (state.animated_cards.empty()) {
    state.animated_cards = state.cards;
  }

  // Animate card positions toward their targets.
  // dt is hardcoded to match the original Python behavior (frame-rate
  // independent lerp was not intended).
  animate(state.animated_cards, state, 0.1f);

  // Draw stack placeholder when a drag is in progress.
  if (state.drag_state.current_stack != -1) {
    draw_stack_placeholder(state.stacks[state.drag_state.current_stack]);
  }

  // Collect stacks and sort by depth (ascending) before drawing.
  std::vector<Stack> stacks_sorted = state.stacks;
  std::sort(
    stacks_sorted.begin(),
    stacks_sorted.end(),
    [](const Stack& a, const Stack& b) { return a.depth < b.depth; }
  );

  // Draw stacks (skip the card that is currently being dragged).
  for (const Stack& stack : stacks_sorted) {
    for (int card_id : stack.cards) {
      if (state.drag_state.card_id == card_id) continue;
      draw_card(state.animated_cards[card_id], stack.face_up);
    }
  }

  // Draw loose cards (always face up).
  for (int card_id : state.loose_cards) {
    draw_card(state.animated_cards[card_id], true);
  }

  // Draw the dragged card on top of everything else.
  if (state.drag_state.card_id >= 0) {
    const KT_Card& card    = state.animated_cards[state.drag_state.card_id];
    bool        face_up = state.stacks[state.drag_state.original_stack].face_up;
    draw_card(card, face_up);
  }

  // Call the optional draw callback for custom overlays.
  if (state.draw_callback) {
    state.draw_callback(&state);
  }

  // Draw zoomed card on top of everything.
  if (state.zoomed_card_id >= 0) {
    // Determine face_up by checking which stack owns this card.
    bool face_up = true;
    for (const Stack& stack : state.stacks) {
      for (int cid : stack.cards) {
        if (cid == state.zoomed_card_id) {
          face_up = stack.face_up;
        }
      }
    }
    draw_zoomed_card(state.cards[state.zoomed_card_id], face_up);
  }
}
